### PR TITLE
Refresh cert as needed

### PR DESCRIPTION
**Pull request by paulproteus**
I believe this is ready to merge. It achieves the following:

* When we have either 0 valid certificates, or 1 valid certificate that expires 3 or fewer days from now, then we get a new cert.

Implementation details:

* In the face of multiple certificates being valid, pick one in a reasonable way.

* It outright denies access to all non-SNI clients by giving them an invalid cert. (@jparyani you might find that amusing.)

* It only calls `.listen()` on the HTTPS socket (fd 3) if a key is available.

Out of scope for this pull request (to be handled tomorrow):

* Deal with clock skew.

### CODE REVIEW

**Comment by paulproteus** [comment]
@kentonv any review you can give this will be nice. I will be doing more testing.

File: meteor-bundle-main.js [hunk 1]
```diff
@@ -73,29 +113,150 @@ function monkeypatchHttpAndHttps() {
...
           if ((e.code === 'EACCES') || (e.code == 'ENOENT')) {
-            console.log("Skipping unreadable HTTPS key information file:", metadataFilename);
+            console.log("Skipping unreadable HTTPS key information file while examining:",
+                        metadataFilename);
             continue;
           } else {
-            throw e;
+            // Sometimes the server gives us a 0-byte response,
+            // presumably due to a connection getting reset?.
+            console.error("Got exception reading JSON from file:", metadataFilename, "Dazed and confused, but trying to continue.", e);
```
Comment (kentonv) [inline meteor-bundle-main.js]: style nitpick: please wrap at 100 columns

File: meteor-bundle-main.js [hunk 2]
```diff
@@ -73,29 +113,150 @@ function monkeypatchHttpAndHttps() {
...
+        validCertificates.push(metadata);
+      }
+
+      // Sort by notBefore, ascending.
+      validCertificates.sort(function(cert1, cert2) {
+        return (cert1.notBefore - cert2.notBefore);
+      });
+
+      if (! validCertificates) {
+        // TODO: Be more robust here, though it's not clear how.
```
Comment (kentonv) [inline meteor-bundle-main.js]: Maybe choose a bogus certificate so that it's more obvious to the user that the problem is with certificates?

Comment (zarvox) [inline meteor-bundle-main.js]: A couple options:

- Attempt to rekey
- Print a log suggesting what action the user can do to fix this

File: meteor-bundle-main.js [hunk 3]
```diff
@@ -73,29 +113,150 @@ function monkeypatchHttpAndHttps() {
...
+          result.nextRekeyTime = secondBest.notBefore;
+        } else {
+          result.nextRekeyTime = validCertificates[0].notAfter;
+        }
+
+        console.log("Will switch to", validCertificates[1].keyFilename,
+                    "at time", new Date(result.nextRekeyTime),
+                    "via real-time SNI re-keying system.");
+      } else {
+        console.log("We have only the one key that we know about. We'll need to renew soon.");
```
Comment (kentonv) [inline meteor-bundle-main.js]: Is the strategy here to have two certificates ready at all times? Does that mean someone installing Sandstorm from scratch receives two weeks of certs instead of one? Would it be possible instead to wait until there's a day left on the current certificate, then fetch the next one?

File: meteor-bundle-main.js [hunk 4]
```diff
@@ -73,29 +113,150 @@ function monkeypatchHttpAndHttps() {
...
+      var result = {
+        ca: validCertificates[0].ca,
+        key: fs.readFileSync(validCertificates[0].keyFilename, 'utf-8'),
+        cert: validCertificates[0].cert
+      };
+
+      // Calculate re-key time.
+      //
+      // - If the cert we want to switch to is not valid yet, then we
+      //   re-key at its notBefore time.
```
Comment (kentonv) [inline meteor-bundle-main.js]: Should there be a grace period to account for clock skew? (The client's clock could be off several minutes.)

File: meteor-bundle-main.js [hunk 5]
```diff
@@ -73,29 +113,150 @@ function monkeypatchHttpAndHttps() {
...
+      //
+      // - If the cert we want to switch to *is* already valid, then
+      //   we re-key at the notAfter time of our current cert.
```
Comment (kentonv) [inline meteor-bundle-main.js]: Why not rekey immediately in this case?

File: meteor-bundle-main.js [hunk 6]
```diff
@@ -73,29 +113,150 @@ function monkeypatchHttpAndHttps() {
...
+      // there are valid certificates for us to use. This could be a
+      // problem if BASE_URL is https but we aren't ready.
+
+      // Create a local variable with key information.
+      var sandcatsState = {};
+
+      // Create a global that others can call into if they want to
+      // re-key, such as when `sandcats.js` downloads a new
+      // certificate.
+      global.sandcats = {};
```
Comment (zarvox) [inline meteor-bundle-main.js]: Why a global here, rather than e.g. pre-meteor instantiating a Sandcats object, which carries these functions?

File: meteor-bundle-main.js [hunk 7]
```diff
@@ -73,29 +113,151 @@ function monkeypatchHttpAndHttps() {
...
+            console.log("Skipping unreadable HTTPS key information file while examining:",
+                        metadataFilename);
             continue;
           } else {
-            throw e;
+            // Sometimes the server gives us a 0-byte response,
+            // presumably due to a connection getting reset?.
+            console.error("Got exception reading JSON from file:", metadataFilename,
+                          "Dazed and confused, but trying to continue.", e);
+            continue;
```
Comment (paulproteus) [inline meteor-bundle-main.js]: Addressed style nitpick (requesting reflow of very long line onto two lines).

File: meteor-bundle-main.js (hunk 2 above)
Comment (paulproteus) [inline meteor-bundle-main.js]: I partly like the idea of choosing a bogus certificate so it's more obvious the problem is with certificates, but I don't like that this would result in users seeing a "OH NO YOUR BROWSER DETECTED SOMETHING TERRIBLE" message.

I think since re-eying only takes ~30 seconds, what I'd want to do is here is:

* Log a message indicating what's up, and
* Don't listen on port 443 yet, and
* Trigger re-keying, and
* When re-keying succeeds, restart the Sandstorm shell by doing `process.exit(99)` and allowing the auto restart process to restart the Sandstorm shell.

I want to restart at the end because the alternative is to provide more and more monkey-patching and proxy object-ing inside this monkey-patched fakeCreateServer() and I don't know that I can confidently do that reliably. It would require not just a mock `createServer` but a mock `HTTP.Server` instance that Meteor will register callbacks on, which seems a bit terrifying.

Possibly someone should tell me to use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy instead (so long as we have those in the version of Meteor we're using).

But note that having 0 key files is a pretty uncommon situation -- anyone who uses the install script will have key files generated during install. This is just for people switching HTTPS on after the fact. So I think that a lot of careful engineering effort in an unlikely codepath is not clearly a good idea.

File: meteor-bundle-main.js (hunk 3 above)
Comment (paulproteus) [inline meteor-bundle-main.js]: The current strategy here is to have two certificates ready at all times. This is for two reasons: (1) It made implementing things easier and (2) if the Sandcats service is (somehow) down for a while, or super buggy, or people turn off their machine and then turn it back on a week later or something like that, I don't want people's certs to expire. (Of course, it _will_ auto re-key on startup, so maybe I could worry about that less.)

Alternate strategy 1: If this is your first cert, get just 1 week, after that, get 2 weeks

* Day 0: new install: have a duration of 1 week (first cert)

* Day 6: Fetch the next week (second cert)

* Day 7: Switch to second cert

* Day 7: Fetch a week after that (third cert)

Alternate strategy 2:

* Always do the renewal one day before the cert is going to expire. This scares me because it gives us a smaller window to deal with sandcats service bugs.

@kentonv do you have a preference for "Alternate strategy 1" or "Alternate strategy 2"?

File: meteor-bundle-main.js (hunk 4 above)
Comment (paulproteus) [inline meteor-bundle-main.js]: Yeah. I was trying to think through the best algorithm here, and wanted to deal with that after submitting this basic implementation.

So thinking aloud:

* Scenario:
    * Sandstorm is already running, and
    * cert 1 expires tomorrow 12:00 PM
    * cert 2 starts today 12:00 PM
    * time right now is 12:00 PM

* Clock that is slow - thinks it's 11:55 AM
    * We keep using cert 1 until "12:00 PM" which is actually 12:05 PM
    * We switch to cert 2 at "12:00 PM" actually 12:05 PM
    * Result: This is OK (because cert 1 didn't expire yet).

* Clock that is fast - thinks it's 12:05 PM
    * We stopped using cert 1 at "12:00 PM" which was actually "11:55 AM"
    * We switched to cert 2 at "12:00 PM" actually 11:55 AM
    * Result: For five minutes, web clients see an expired certificate.
    * Proposal to fix this: Do the switch-over at the _average_ of the cert 1 expiry & cert 2 start date.
    * Alternate proposal: Do the switch-over some constant factor into cert 2 (e.g. 10 minutes) so long as cert 1 is still valid for that time.

File: meteor-bundle-main.js (hunk 6 above)
Comment (paulproteus) [inline meteor-bundle-main.js]: Ya, I'm using a global here because I'm at the most top-level file, before Meteor exists yet, and I want to provide access to some state in this file. To try to be clean, I do it by proving a function that encapsulates the important aspect of state, rather than providing the raw state.

Other ideas welcome.

File: meteor-bundle-main.js (hunk 4 above)
Comment (paulproteus) [inline meteor-bundle-main.js]: * Scenario:
    * Sandstorm is starting up right now, and
    * cert 1 expires tomorrow 12:00 PM
    * cert 2 starts today 12:00 PM
    * time right now is 11:58 AM

* Clock that is 5 min slow - thinks it's 11:53 AM
    * We keep using cert 1 until "12:00 PM" which is actually 12:05 PM
    * We switch to cert 2 at "12:00 PM" which is actually 12:05 PM
    * Result: This is OK (because cert 1 didn't expire yet).

* Clock that is 5 minutes fast - thinks it's 12:03 PM
    * We'll pick cert 1 since it's still valid and we prefer older certs that are still valid.
    * We will tell ourselves to re-key at tomorrow "12:00 PM" (actually 11:55 AM)
    * Result: This is OK (because we ditched cert 1 a few min early but that was OK).

File: meteor-bundle-main.js (hunk 5 above)
Comment (paulproteus) [inline meteor-bundle-main.js]: Doing it this way maximizes how long we use each cert, but I'll probably rethink this as part of the "only renew four days before expiry" goal.

File: meteor-bundle-main.js (hunk 3 above)
Comment (paulproteus) [inline meteor-bundle-main.js]: Per discussion Kenton prefers something more like "Alternate Strategy 2" where we pick a 2-3 day window. I'll go with a 4 day window.

File: meteor-bundle-main.js (hunk 5 above)
Comment (paulproteus) [inline meteor-bundle-main.js]: It used to be the case that we calculated the answer to "Should I request a fresh cert from sandcats.io?" based on the rekey time. But now we don't do that. So I guess we can use the newest cert that is still valid if we want.

Comment (paulproteus) [inline meteor-bundle-main.js]: It'd be nice if we could always rekey immediately, but I prefer my current algorithm because it's easier for me to explain to myself how it operates in the face of clock skew. Can think about this more, but would prefer to do so on a separate pull request if the current version is not wrong.

File: meteor-bundle-main.js (hunk 4 above)
Comment (paulproteus) [inline meteor-bundle-main.js]: Will handle clock skew in a separate pull request, if that's OK.
